### ISSUE TITLE
Front End: SPA vs MPA

### ISSUE THREAD

**Issue by pelegyair** [TechnionYearlyProject/DailyPulseMe#33]
this issue will discuss which method is better of our site, single page web applications(SPA) vs multi page web applications(MPA).

**MPA** is a traditional kind of web application. This means that every time the application needs to display the data or submit data back to server it has to request a new page from the server and then render it in the web browser. Nothing is wrong with this approach for simple applications. But if there is a need to create a rich user interface then the page might become very complex and to be loaded with a lot of data. Generating complex pages on a server and transfer them to the client over the internet and rendering them into the browser takes time and degrade the user experience because of that. In the beginning of 2000s MPA was improved by introducing AJAX, which allowed refreshing only parts of the page and not the whole page. It helped to improve the user experience but added complexity to the page itself.

Ten years later the **SPA** pattern emerged. SPA is essentially an evolution of the MPA+AJAX design pattern, where the only shell page is generated on the server and all UI is rendered by browser JavaScript code. SPA requests the markup and data separately, and renders pages directly in browser. We can do this because of the new advanced MVVM JavaScript frameworks emerged like AngularJS and KnockoutJS.

**SPA advantages over MPA:**

- Faster page loading times
- Improved user experience because the data is loading in the background form server
- No need to write the code to render pages on the server
- Decoupling of front-end and back-end development
- Simplified mobile development; you can reuse the same backend for web application and native mobile application

**SPA disadvantages to MPA:**

- Heavy client frameworks which are required to be loaded to the client
- UI code is not compiled, so it’s harder to debug and it’s exposed to potential malicious user
- SEO (search engine optimization)  implications; since your pages are built in the browser, the search engine crawler will see a different version of the page than that of your users

many articles and sites suggest the hybrid way, a few pages that will implement as SPA. this method might sound good but it includes the cons of both of the ways. in addition, i dont think that our site needs this kind of "power".
In conclusion i believe that the right way to build our web app is **SPA**.

**to understand this topic i used:
http://www.eikospartners.com/blog/multi-page-web-applications-vs.-single-page-web-applications
and
https://medium.com/@NeotericEU/single-page-application-vs-multiple-page-application-2591588efe58

* i believe that this issue is important because understanding it helps us choose the right java framework to work with.
